### PR TITLE
sql: support `CREATE INDEX ... WITH`

### LINKED ISSUE #4231: CREATE INDEX ... WITH
Via #3636 we got the awesome ability to set the compaction window per index. Unfortunately this is only available via `ALTER INDEX ... SET` after creating the index—it should be possible to say `CREATE INDEX ... WITH (logical_compaction_window = ...)` to do it all one shot.

/cc @ruchirK

### PR DESCRIPTION
**Pull request by benesch**
Add support for setting index options directly at index creation time.
Previously setting these options required a separate call to `ALTER
INDEX`.

Fix #4231.

---
This change is [Reviewable](https://reviewable.io/reviews/materializeinc/materialize/4263)

### CODE REVIEW

File: src/sql/src/plan/statement.rs [hunk 1]
```diff
@@ -480,7 +480,7 @@ fn plan_index_options(options: &[SqlOption]) -> Result<Vec<IndexOption>, anyhow:
         Some(Value::String(window)) => {
             let window = match window.as_str() {
                 "off" => None,
-                s => Some(parse_duration::parse(s)?),
+                s => Some(strconv::parse_interval(s)?.duration()?),
```
Comment (ruchirK) [inline src/sql/src/plan/statement.rs:483]: any particular reason you didn't use `repr::util::parse_duration` here?

Comment (benesch) [reply src/sql/src/plan/statement.rs:483]: 🤦‍♂️ thanks. This is the call site that _inspired_ that helper method and I completely forgot to use it here.

File: src/repr/src/lib.rs [hunk 2]
```diff
@@ -28,6 +28,7 @@ mod scalar;
 
 pub mod adt;
 pub mod strconv;
+pub mod util;
```
Comment (ruchirK) [inline src/repr/src/lib.rs:31]: [minor nit / rant; feel free to ignore]

Idk how you feel about this but imo `util` crates / files end up one of two ways after a year or two

a) same size as the original
b) accrued a bunch of random stuff, now unclear what all is in `util` and its a source of tech debt

I don't have a good suggestion for a different name (maybe `arg_duration`?) but would ask you to consider alternatives.

File: src/sql/src/plan/statement.rs [hunk 3]
```diff
@@ -435,65 +435,65 @@ fn handle_alter_index_options(
...
         Err(e) => return Err(e.into()),
     };
 
-    Ok(Plan::AlterIndexLogicalCompactionWindow(alter_index))
+    match options {
+        AlterIndexOptionsList::Reset(options) => {
+            let options = options
+                .into_iter()
+                .filter_map(|o| match normalize::ident(o).as_str() {
+                    "logical_compaction_window" => Some(IndexOptionName::LogicalCompactionWindow),
```
Comment (ruchirK) [inline src/sql/src/plan/statement.rs:461]: nice! i like this much better :)

File: src/coord/src/coord.rs [hunk 4]
```diff
@@ -2431,6 +2430,18 @@ where
         Ok(())
     }
 
+    fn set_index_options(&mut self, id: GlobalId, options: Vec<IndexOption>) {
+        let index = self.indexes.get_mut(&id).expect("index known to exist");
```
Comment (ruchirK) [inline src/coord/src/coord.rs:2434]: my reasoning earlier was that someone could interleave index creation / alteration and delete in such a way that we might get to this point and have the index be deleted. is that not actually possible? (don't totally remember my thought process offhand, not taking the time to read through the file rn)

**Comment by ruchirK** [comment]
Looks great to me! Thanks for knocking this one out!

File: src/coord/src/coord.rs (hunk 4 above)
Comment (benesch) [reply src/coord/src/coord.rs:2434]: Indeed it is not! The correctness of the whole system hinges on the fact that the coordinator is single threaded and always applies the effects of a SQL query immediately after planning it. So the read-only copy of the catalog that the SQL planner uses to compute the plan is guaranteed to be the same catalog that the coordinator is applying updates to.

Comment (ruchirK) [reply src/coord/src/coord.rs:2434]: ok i remember now i was going off of sinks -- but sinks have the asynchronous step where they create external state that breaks the order of updates guarantee

**Comment by ruchirK** [comment]
Is there anything I can do to help push these changes out?

**Comment by benesch** [comment]
Unfortunately not. This is blocked on getting support for the `millisecond`
unit in time intervals.

**Comment by benesch** [comment]
Ok, I'm ditching the commit that tries to ditch `parse_duration`, as I think the rest is mergeable. Will help with #5335.

**Pull request merged**